Fix nested functions ignoring prefixed variable assignments

JohnoKing opened the pull request:
This pull request fixes the bug described in att/ast#32. The following explanation is from att/ast#467:

> While copying variables from function's local scope to a new scope, variable attributes were not copied. Such variables were not marked to be exported in the new function. For e.g.
> ```sh
> function f2 { env | grep -i "^foo"; }
> function f1 { env | grep -i "^foo"; f2; }
> foo=bar f1
> ```
> 
> prints 'foo=bar' only once, but it should print be twice.

The fix is to copy the flags a variable has in the parent function's local scope to the nested function's scope.

McDutchie commented on src/cmd/ksh93/tests/functions.sh:
```diff
@@ -1258,5 +1258,12 @@ $SHELL -c 'PATH=/dev/null; fn() { unset -f fn; true; }; fn; fn' 2> /dev/null
 
+# ======
+# Check if environment variables passed while invoking a function are exported
+# https://github.com/att/ast/issues/32
```
Please `unset foo` here, just to be sure…

JohnoKing committed:
Fix nested functions ignoring prefixed variable assignments

This commit fixes the bug described in att/ast#32. The fix and
following explanation is from att/ast#467:

While copying variables from function's local scope to a new scope,
variable attributes were not copied. Such variables were not marked
to be exported in the new function. For e.g.

function f2 { env | grep -i "^foo"; }
function f1 { env | grep -i "^foo"; f2; }
foo=bar f1

prints 'foo=bar' only once, but it should print be twice.

src/cmd/ksh93/sh/xec.c:
 - When variables from the local scope of a function are copied into
   the scope of a nested function, the attributes of the variables
   need to be copied as well.

src/cmd/ksh93/tests/functions.sh:
 - Add regression tests from ksh2020 to check environment variables
   passed to functions.
